src/conn/pool/mod.rs

```diff
@@ -338,6 +338,75 @@ impl Pool {
         self.poll_new_conn_inner(false)
     }
 
+    /// Try to acquire a new connection under a spinlock checking that we don't
+    /// violate `exist < self.pool_constraints.max()`.
+    fn spin_acquire_conn(&mut self) -> Option<Result<Async<GetConn>>> {
+        let mut exist = self.inner.exist.load(atomic::Ordering::Acquire);
+
+        // Try to atomically swap the exist value while checking our condition.
+        loop {
+            if exist >= self.pool_constraints.max() {
+                return None;
+            }
+
+            let swapped =
+                self.inner
+                    .exist
+                    .compare_and_swap(exist, exist + 1, atomic::Ordering::AcqRel);
```

Comment:
I'm pretty sad to see this become a `compare_and_swap`, as that scales a fair bit worse than `fetch_add` does. In theory, only as many tasks as can proceed should be woken up, so I think being optimistic about being given _a_ slot here (as `fetch_add` does) might be the way to go.

Reply:
It might also be that we can replace this whole loop with https://doc.rust-lang.org/std/sync/atomic/struct.AtomicUsize.html#method.fetch_update one day :)

Reply:
Also note that `fetch_add` would allow multiple clients to proceed concurrently if multiple slots are freed up at the same time (which I suspect will be somewhat common).

Reply:
Just got back form a nasty cold. Sorry for the late response.

I think your points are valid from a performance perspective. But if we really wanna dig into it we'd have to benchmark in real world scenarios to see how much of a difference it makes. If you don't think this proposal clarifies the implementation enough to hold its own weight I'm fine with just closing it for now.